internal/app/loki-multi-tenant-proxy/auth.go

```diff
@@ -3,49 +3,137 @@ package proxy
 import (
 	"context"
 	"crypto/subtle"
+	b64 "encoding/base64"
+	"encoding/json"
+	"fmt"
 	"net/http"
+	"strings"
 
+	"github.com/coreos/go-oidc"
 	"github.com/giantswarm/loki-multi-tenant-proxy/internal/pkg"
+	"go.uber.org/zap"
 )
 
 type key int
 
+// Struct to represent the OAuth token payload section
+type Payload struct {
+	Iss           string   `json:"iss"`
+	Sub           string   `json:"sub"`
+	Aud           string   `json:"aud"`
+	Exp           int      `json:"exp"`
+	Iat           int      `json:"iat"`
+	AtHash        string   `json:"at_hash"`
+	CHash         string   `json:"c_hash"`
+	Email         string   `json:"email"`
+	EmailVerified bool     `json:"email_verified"`
+	Groups        []string `json:"groups"`
+	Name          string   `json:"name"`
+}
+
 const (
 	// OrgIDKey Key used to pass loki tenant id though the middleware context
 	OrgIDKey key = iota
 	realm        = "Loki multi-tenant proxy"
+	readUser     = "read"
 )
 
-// BasicAuth can be used as a middleware chain to authenticate users before proxying a request
-func BasicAuth(handler http.HandlerFunc, authConfig *pkg.Authn) http.HandlerFunc {
+func Authentication(handler http.HandlerFunc, authConfig *pkg.Authn, logger *zap.Logger) http.HandlerFunc {
 	return func(w http.ResponseWriter, r *http.Request) {
-		user, pass, ok := r.BasicAuth()
-		authorized, orgID := isAuthorized(user, pass, authConfig)
-		if !ok || !authorized {
-			writeUnauthorisedResponse(w)
+		token, ok := r.Header["X-Id-Token"]
+		var authenticationMode string
+		var user = "read" // default user if oauth mode
+		var pass = ""     // no password in oauth mode
+		if ok {
+			// OAuth token authentication mode (X-Id-Token header provided)
+			logger.Info("OAuth authentication mode")
+			logger.Info(fmt.Sprintf("Token = %s", token[0]))
+			// Decode OAuth token payload section
+			payload, err := decodeOAuthToken(token[0])
+			if err != nil {
+				logger.Error(fmt.Sprintf("Error decoding token payload %s", token[0]), zap.Error(err))
+				return
+			}
+			// Token validation against Dex
+			err = validateOAuthToken(token[0], payload.Iss, payload.Aud, r.Context())
+			if err != nil {
+				logger.Error(fmt.Sprintf("Error while validating OAuth token against DEX %s", token[0]), zap.Error(err))
```

Comment:
I don't think there's anything dex-specific here.
```suggestion
				logger.Error(fmt.Sprintf("Error while validating OAuth token against identity provider %s", token[0]), zap.Error(err))
```